yarpl/include/yarpl/flowable/FlowableOperator.h

```diff
@@ -55,48 +55,110 @@ class FlowableOperator : public Flowable<D> {
     }
 
     void subscriberOnNext(D value) {
-      subscriber_->onNext(std::move(value));
-    }
-
-    void request(int64_t delta) override {
-      upstream_->request(delta);
+      if (subscriber_) {
+        subscriber_->onNext(std::move(value));
+      }
     }
 
-    // this method should be used to terminate the operators
+    /// Terminates both ends of an operator normally.
     void terminate() {
-      subscriber_->onComplete(); // should break the cycle to this
-      upstream_->cancel(); // should break the cycle to this
+      if (isTerminated()) {
+        return;
+      }
+
+      if (auto upstream = std::move(upstream_)) {
+        upstream->cancel();
+      }
+      if (auto subscriber = std::move(subscriber_)) {
+        subscriber->onComplete();
+      }
+
       Refcounted::decRef(*this);
     }
 
-    void cancel() override {
-      assert(upstream_ && "subscription was already terminated");
-      (decltype(upstream_)(std::move(upstream_)))->cancel();
-      subscriber_.reset(); // breaking the cycle
+    /// Terminates both ends of an operator with an error.
+    void terminateErr(std::exception_ptr eptr) {
+      if (isTerminated()) {
+        return;
+      }
+
+      if (auto upstream = std::move(upstream_)) {
+        upstream->cancel();
+      }
+      if (auto subscriber = std::move(subscriber_)) {
+        subscriber->onError(std::move(eptr));
+      }
+
       Refcounted::decRef(*this);
     }
 
-    void onComplete() override {
-      assert(upstream_ && "subscription was already terminated");
-      upstream_.reset(); // breaking the cycle
-      subscriber_->onComplete();
-      Refcounted::decRef(*this);
+    // Subscription.
+
+    void request(int64_t delta) override {
+      if (upstream_) {
+        upstream_->request(delta);
+      }
     }
 
-    void onError(std::exception_ptr error) override {
-      assert(upstream_ && "subscription was already terminated");
-      upstream_.reset(); // breaking the cycle
-      subscriber_->onError(error);
+    void cancel() override {
+      if (isTerminated()) {
+        return;
+      }
+
+      if (auto upstream = std::move(upstream_)) {
+        upstream->cancel();
+      }
+
+      subscriber_.reset();
+
       Refcounted::decRef(*this);
     }
 
+    // Subscriber.
+
     void onSubscribe(
         Reference<yarpl::flowable::Subscription> subscription) override {
+      if (upstream_) {
+        subscription->cancel();
+        return;
+      }
+
       upstream_ = std::move(subscription);
       subscriber_->onSubscribe(Reference<yarpl::flowable::Subscription>(this));
     }
 
+    void onComplete() override {
+      if (isTerminated()) {
+        return;
+      }
+
+      upstream_.reset();
+
+      if (auto subscriber = std::move(subscriber_)) {
+        subscriber->onComplete();
+      }
+
+      Refcounted::decRef(*this);
+    }
+
+    void onError(std::exception_ptr eptr) override {
```

Comment:
It isn't.  `terminateErr()` cancels the upstream whereas this does not.  The different comes from which part of the pipeline we're getting a termination signal.  `onError()` means the upstream has sent the termination, so we only need to null out the `upstream_` reference.  `terminateErr()` is used when the operator itself wants to terminate, so we need to send a `cancel()` to the upstream (and an `onError()` to the downstream as well).